datafusion/core/src/physical_plan/sorts/merge.rs

```diff
@@ -210,7 +218,14 @@ impl<C: Cursor> SortPreservingMergeStream<C> {
                 self.loser_tree_adjusted = false;
                 self.in_progress.push_row(stream_idx);
                 if self.in_progress.len() < self.batch_size {
-                    continue;
+                    match self.fetch {
```

Comment:
I feel like `self.in_progress` should initialize with `fetch` length, if `fetch` is smaller than `batch_size`. Need to think about it
https://github.com/apache/arrow-datafusion/blob/b12e4546392e5dbb8a355b61373a3e1fc0c3f92d/datafusion/core/src/physical_plan/sorts/merge.rs#L144

Reply:
> The loser tree, because it provides the loser of each match, will contain repeat nodes. Since the heap is a data-storing structure, it won't contain these redundancies. Another difference between the two is that the loser tree must be a full binary tree (because it is a type of tournament tree), but the heap does not necessarily have to be binary
> https://stackoverflow.com/questions/7685943/whats-the-difference-between-a-heap-and-a-loser-tree-in-external-sorting

based on the above, do we need to switch to Heap to reduce memory usage?